### PR TITLE
[Local] Only intersect llvm.access.group metadata if instr moves.

### PR DESCRIPTION
**Pull request by fhahn**
Preserve llvm.access.group metadata on the replacement instruction, if it does not move. In that case, the program would be UB, if the parallel property encoded in the metadata does not hold.

### CODE REVIEW

**Review by nikic — COMMENTED** [review]
Before we start relying on this in one way or the other, I'd like to have it spelled out that this is UB (rather than poison) in LangRef, so we don't have to guess next time.

**Comment by fhahn** [comment]
> Before we start relying on this in one way or the other, I'd like to have it spelled out that this is UB (rather than poison) in LangRef, so we don't have to guess next time.

Tried to do so in https://github.com/llvm/llvm-project/pull/116220

File: llvm/test/Transforms/InstCombine/loadstore-metadata.ll [hunk 1]
```diff
@@ -261,8 +261,8 @@ entry:
 define double @preserve_load_metadata_after_select_transform_metadata_missing_3(ptr %a, ptr %b) {
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/loadstore-metadata.ll:261]: Drop FIXME?

By the way, I think these tests would be clearer if you didn't have equal metadata on the first two loads, it's a bit unclear how exactly the preservation works that way.

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/test/Transforms/InstCombine/loadstore-metadata.ll (hunk 1 above)
Comment (fhahn) [reply llvm/test/Transforms/InstCombine/loadstore-metadata.ll:258]: Dropped and added a new variant in 9e0ea8c8816d9f5837ad3357be32dddce28c7fb1

Comment (fhahn) [reply llvm/test/Transforms/InstCombine/loadstore-metadata.ll:258]: made distinct....

**Pull request merged**